sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/FluxTrace.java

```diff
@@ -56,23 +55,17 @@ protected void hookOnSubscribe(Subscription subscription) {
 
         @Override
         protected void hookOnNext(ServiceBusMessageContext message) {
-            Throwable exception = null;
             Context span = instrumentation.instrumentProcess("ServiceBus.process", message.getMessage(), Context.NONE);
 
-            AutoCloseable scope  = tracer.makeSpanCurrent(span);
+            AutoCloseable scope = tracer.makeSpanCurrent(span);
             try {
                 downstream.onNext(message);
-            } catch (Throwable t) {
-                exception = t;
-            } finally {
-                Context context = message.getMessage().getContext();
-                if (context != null) {
-                    Object processorException = context.getData(PROCESS_ERROR_KEY).orElse(null);
-                    if (processorException instanceof Throwable) {
-                        exception = (Throwable) processorException;
-                    }
+                if (!instrumentation.isProcessorInstrumentation()) {
+                    tracer.endSpan(null, span, scope);
```

Comment:
yes, flux trace here is the first to be invoked after message is received and when it's dispatched to the app - we track metrics and start span there for async receiver and processor (sync receiver is a different beast).

In case of async client, we have limited control over subscriber behavior and don't know if it's sync or async - the best we can do is to end span here once downstream subscriber `onNext` completes.

In the case of processor, we use parallel flux there and the downstream subscriber completes before the processing callback starts.  So instead of ending the span here, we keep it in the context attached to the message (it was well-hidden  - I made it a bit more straightforward).

In the processor, we'll get this span from message and end it there.

The order of things in case of consurrent processing is the following:
1. message is received and is being dispatched to the user's subscriber
2. FluxTrace starts processing span and attaches it to the message's context
3. downstream subscribers do their stuff (lock renewal, messages are added to internal queue for parallel scheduling)
4. `downstream.onNext(message)` completes
5. process callback on processor is called
6. once it's over, we get span from the message context and end it

I hope to find a nicer/easier way to plug tracing in the v2 stack.

Reply:
Thank you for the clear and detailed explanation! Yes, this is easy to plug into v2 stack. Once this is improvement is merged, I will try to replicate this in V2 and request your feedback.